src/Language/PureScript/PSString.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Language.PureScript.PSString (PSString, mkString, renderJSON, toUTF16CodeUnits, codePoints) where
+
+import Prelude.Compat
+import Numeric (showHex)
+import Data.List (unfoldr)
+import Data.Monoid ((<>))
+import Data.String (IsString(..))
+import Data.Text (Text)
+import Data.Word (Word16)
+import qualified Data.Aeson as A
+import qualified Data.Text as T
+
+-- |
+-- Strings in PureScript are sequences of UTF-16 code units, which do not
+-- necessarily represent UTF-16 encoded text. For example, it is permissible
+-- for a string to contain *lone surrogates,* i.e. characters in the range
+-- U+D800 to U+DFFF which do not appear as a part of a surrogate pair.
+--
+--
+newtype PSString = PSString [Word16]
+  deriving (Eq, Ord, A.FromJSON, Monoid)
+
+instance Show PSString where
+  show = show . codePoints
+
+-- TODO: this shouldn't have to be exposed; try to replace its usages with more specialised functions
+codePoints :: PSString -> String
+codePoints = unfoldr decode . toUTF16CodeUnits
+  where
+  decode :: [Word16] -> Maybe (Char, [Word16])
+  decode (h:l:rest) | h >= 0xD800 && h <= 0xDBFF && l >= 0xDC00 && l <= 0xDFFF = Just (unsurrogate h l, rest)
+  decode (c:rest) = Just (toChar c, rest)
+  decode [] = Nothing
+
+  unsurrogate :: Word16 -> Word16 -> Char
+  unsurrogate h l = toEnum ((toInt h - 0xD800) * 0x400 + (toInt l - 0xDC00) + 0x10000)
+
+instance IsString PSString where
+  fromString a = PSString $ concatMap encodeUTF16 a
+    where
+    surrogates :: Char -> (Word16, Word16)
+    surrogates c = (toWord (h + 0xD800), toWord (l + 0xDC00))
+      where (h, l) = divMod (fromEnum c - 0x10000) 0x400
+
+    encodeUTF16 :: Char -> [Word16]
+    encodeUTF16 c | fromEnum c > 0xFFFF = [high, low]
+      where (high, low) = surrogates c
+    encodeUTF16 c = [toWord $ fromEnum c]
+
+instance A.ToJSON PSString where
+  toJSON = A.toJSON . map toChar . toUTF16CodeUnits
+
+renderJSON :: PSString -> Text
+renderJSON s = T.pack $ "\"" <> concatMap encodeChar (toUTF16CodeUnits s) <> "\""
```

Comment:
Unfortunately, `T.concatMap` takes a `Text` as its second parameter, so I'll have to keep this the way it is.